### ISSUE TITLE
Trying to open a particular Realm DB file throws the following exception - "Upgrading from file format version 22 to 24 D:\a\realm-dotnet\realm-dotnet\wrappers\realm-core\src\realm/util/encrypted_file_mapping.hpp:180: [realm-core-14.5.1] Assertion failed: local_ndx < m_page_state.size() with (local_ndx, m_page_state.size(), size_t(addr), size_t(m_addr), m_page_shift) =  [85, 1, 2331762327552, 2331762294784, 16] <backtrace not supported on this platform>"

### ISSUE THREAD

**Issue by clement-ilango** [realm/realm-dotnet#3592]
### What happened?

Info: DB: 61583 Thread 48384: Upgrading from file format version 22 to 24
D:\a\realm-dotnet\realm-dotnet\wrappers\realm-core\src\realm/util/encrypted_file_mapping.hpp:180: [realm-core-14.5.1] Assertion failed: local_ndx < m_page_state.size() with (local_ndx, m_page_state.size(), size_t(addr), size_t(m_addr), m_page_shift) =  [85, 1, 2331762327552, 2331762294784, 16]
<backtrace not supported on this platform>

### Repro steps

Same as above.

### Version

.NET FRAMEWORK 4.7.2

### What Atlas Services are you using?

Local Database only

### What type of application is this?

WPF

### Client OS and version

Windows 10 

### Code snippets

            var config = new RealmConfiguration(fileName);
            config.SchemaVersion = 6;

            config.Schema = new[] {....}

            if (!string.IsNullOrEmpty(FileName) && File.Exists(FileName))
            {
                var hex = File.ReadAllText(FileName);
                byte[] bytes = Misc.HexToBytes(hex);
                config.Key = bytes;
            }
            else
            {
                var hex = "";
                if (Settings.Instance.AndroidMobile.KeyEnabledOrNot)
                {
                    hex = Settings.Instance.AndroidMobile.Key;
                    if (string.IsNullOrEmpty(hex))
                        throw new Exception(Resources.strPleaseProvideKey);
                }
                if (!string.IsNullOrEmpty(hex))
                {
                    byte[] bytes = Misc.HexToBytes(hex);
                    config.Key = bytes;
                }
            }

            using (var realm = Realm.GetInstance(config))
            {
                  //Do some things with the opened Realm DB
            }

### Relevant log output

```shell
2024-05-01 23:47:31.993 Info: DB: 61583 Thread 48384: Upgrading from file format version 22 to 24
D:\a\realm-dotnet\realm-dotnet\wrappers\realm-core\src\realm/util/encrypted_file_mapping.hpp:180: [realm-core-14.5.1] Assertion failed: local_ndx < m_page_state.size() with (local_ndx, m_page_state.size(), size_t(addr), size_t(m_addr), m_page_shift) =  [85, 1, 2331762327552, 2331762294784, 16]
<backtrace not supported on this platform>

**Comment by nirinchev** [comment]
You haven't specified the version of the Realm SDK that you're using, but this appears to be a bug that was fixed in the latest version - can you try upgrading to confirm?

**Comment by clement-ilango** [comment]
Hi there,
It happens with the latest version of your SDK too. I can send you the realm DB file and it’s encryption key if you need it. Realm Studio is able to open that file but even that is not able to show me the tables that are inside that Realm DB file. It seems to believe there are no tables in it for whatever reason. Realm Studio 11.2.1 is able to open and list it’s table contents fine (which is what we were using till now and Realm SDK 10.7.1 I believe). The mobile team upgraded their Realm library and I have to make sure the tools I am working on are able to work with their realm files (that are produced now) and that their prior versions of their app created (backward computability which is where this Bug report is coming from).
Thanks,
Clement.
P.S:- Attached is the realm file (and it’s backup Realm Studio made). The DB encryption key is 792cddc553a0452ee893e7583735d4e9f0942d6c8404398f612fd5e415e115e19443aa3be112e072f5fc22b7e2471a11ad2924c19413d84d19c32179c1063bd2.
***@***.***


'RealmDBextractorEX.exe' (CLR v4.0.30319: DefaultDomain): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_64\mscorlib\v4.0_4.0.0.0__b77a5c561934e089\mscorlib.dll'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: DefaultDomain): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\RealmDBextractorEX.exe'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\PresentationFramework\v4.0_4.0.0.0__31bf3856ad364e35\PresentationFramework.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\WindowsBase\v4.0_4.0.0.0__31bf3856ad364e35\WindowsBase.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Core\v4.0_4.0.0.0__b77a5c561934e089\System.Core.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System\v4.0_4.0.0.0__b77a5c561934e089\System.dll'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_64\PresentationCore\v4.0_4.0.0.0__31bf3856ad364e35\PresentationCore.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Xaml\v4.0_4.0.0.0__b77a5c561934e089\System.Xaml.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\PrivateAssemblies\Runtime\Microsoft.VisualStudio.Debugger.Runtime.Desktop.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Configuration\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Configuration.dll'. Symbols loaded without source information.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Xml\v4.0_4.0.0.0__b77a5c561934e089\System.Xml.dll'. Symbols loaded without source information.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\log4net.dll'. Module was built without symbols.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\SharedResources.dll'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\PresentationFramework.Aero2\v4.0_4.0.0.0__31bf3856ad364e35\PresentationFramework.Aero2.dll'. Symbols loaded without source information.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\PresentationFramework-SystemXml\v4.0_4.0.0.0__b77a5c561934e089\PresentationFramework-SystemXml.dll'. Symbols loaded without source information.
Arguments: shell pm list packages
StdErr: adb.exe: no devices/emulators found

StdOut: 0 bytes
StdOut: adb.exe: no devices/emulators found

'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'c:\program files\microsoft visual studio\2022\professional\common7\ide\commonextensions\microsoft\xamldiagnostics\Framework\x64\Microsoft.VisualStudio.DesignTools.WpfTap.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.Serialization\v4.0_4.0.0.0__b77a5c561934e089\System.Runtime.Serialization.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\SMDiagnostics\v4.0_4.0.0.0__b77a5c561934e089\SMDiagnostics.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.ServiceModel.Internals\v4.0_4.0.0.0__31bf3856ad364e35\System.ServiceModel.Internals.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\UIAutomationTypes\v4.0_4.0.0.0__31bf3856ad364e35\UIAutomationTypes.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\UIAutomationProvider\v4.0_4.0.0.0__31bf3856ad364e35\UIAutomationProvider.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Windows.Forms\v4.0_4.0.0.0__b77a5c561934e089\System.Windows.Forms.dll'. Symbols loaded without source information.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Drawing\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Drawing.dll'. Symbols loaded without source information.
Managed Debugging Assistant 'LoadFromContext'
The assembly named 'OptaneShellExt' was loaded from 'file:///C:/windows/System32/DriverStore/FileRepository/iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da/OptaneShellExt.dll' using the LoadFrom context. The use of this context can result in unexpected behavior for serialization, casting and dependency resolution. In almost all cases, it is recommended that the LoadFrom context be avoided. This can be done by installing assemblies in the Global Assembly Cache or in the ApplicationBase directory and using Assembly.Load when explicitly loading assemblies.

'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\System32\DriverStore\FileRepository\iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da\OptaneShellExt.dll'.
Managed Debugging Assistant 'LoadFromContext'
The assembly named 'SharpShell' was loaded from 'file:///C:/windows/System32/DriverStore/FileRepository/iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da/SharpShell.DLL' using the LoadFrom context. The use of this context can result in unexpected behavior for serialization, casting and dependency resolution. In almost all cases, it is recommended that the LoadFrom context be avoided. This can be done by installing assemblies in the Global Assembly Cache or in the ApplicationBase directory and using Assembly.Load when explicitly loading assemblies.

'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\System32\DriverStore\FileRepository\iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da\SharpShell.dll'.
Managed Debugging Assistant 'LoadFromContext'
The assembly named 'log4net' was loaded from 'file:///C:/windows/System32/DriverStore/FileRepository/iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da/log4net.DLL' using the LoadFrom context. The use of this context can result in unexpected behavior for serialization, casting and dependency resolution. In almost all cases, it is recommended that the LoadFrom context be avoided. This can be done by installing assemblies in the Global Assembly Cache or in the ApplicationBase directory and using Assembly.Load when explicitly loading assemblies.

'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\System32\DriverStore\FileRepository\iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da\log4net.dll'.
Managed Debugging Assistant 'LoadFromContext'
The assembly named 'PinningServiceApi' was loaded from 'file:///C:/windows/System32/DriverStore/FileRepository/iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da/PinningServiceApi.DLL' using the LoadFrom context. The use of this context can result in unexpected behavior for serialization, casting and dependency resolution. In almost all cases, it is recommended that the LoadFrom context be avoided. This can be done by installing assemblies in the Global Assembly Cache or in the ApplicationBase directory and using Assembly.Load when explicitly loading assemblies.

'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\System32\DriverStore\FileRepository\iastorpinningcomponent.inf_amd64_5c30a67a1f3ea5da\PinningServiceApi.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.ComponentModel.Composition\v4.0_4.0.0.0__b77a5c561934e089\System.ComponentModel.Composition.dll'.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\Realm.dll'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\netstandard\v4.0_2.0.0.0__cc7b13ffcd2ddd51\netstandard.dll'. Module was built without symbols.
The thread 0xb708 has exited with code 0 (0x0).
The thread 0x11634 has exited with code 0 (0x0).
The thread 0xfc0c has exited with code 0 (0x0).
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.ValueTuple\v4.0_4.0.0.0__cc7b13ffcd2ddd51\System.ValueTuple.dll'. Module was built without symbols.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\windows\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.InteropServices.RuntimeInformation\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Runtime.InteropServices.RuntimeInformation.dll'. Module was built without symbols.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\Realm.PlatformHelpers.dll'. Symbols loaded.
'RealmDBextractorEX.exe' (CLR v4.0.30319: RealmDBextractorEX.exe): Loaded 'C:\Users\ilangcx\source\repos\uuumain\android_app_extractor\Android_App_Extractor\bin\x64\Debug\MongoDB.Bson.dll'.
2024-05-02 18:02:42.373 Info: DB: 17198 Thread 42816: Upgrading from file format version 22 to 24
D:\a\realm-dotnet\realm-dotnet\wrappers\realm-core\src\realm/util/encrypted_file_mapping.hpp:184: [realm-core-14.6.1] Assertion failed: local_ndx < m_page_state.size() with (local_ndx, m_page_state.size(), size_t(addr), size_t(m_addr), m_page_shift) =  [85, 1, 1549425541120, 1549425508352, 16]
<backtrace not supported on this platform>
!!! IMPORTANT: Please report this at https://github.com/realm/realm-core/issues/new/chooseThe program '[0xDCD8] RealmDBextractorEX.exe' has exited with code 3221226505 (0xc0000409).

**Comment by nirinchev** [comment]
@ironage this appears to be similar to the encryption bug that you fixed, but continues to happen even with Core 14.6.1, which should contain the fix. Is there anything obvious from the stacktrace or something we can ask the reporter to try track it down?

**Comment by ironage** [comment]
Hi @clement-ilango the stack traces you provided show you are using a core version of v14.5.1, and the relevant fix is in v14.6.0. For the .Net SDK this is version [12.1.0 (2024-05-01)](https://github.com/realm/realm-dotnet/blob/main/CHANGELOG.md#1210-2024-05-01) or later. Please try upgrading and let us know if this solves the issue. If not, please provide an updated stack trace, and send your Realm file with encryption key to [realm-help@mongodb.com](mailto:realm-help@mongodb.com).

**Comment by clement-ilango** [comment]
Hi James,
Attached are copies of that file that you need. DB key you need to open it is below. You would see that you can open it successfully in latest version of Realm Studio and RealmStudio older versions (of which I use is version 11.2.1) but you would see the tables listed only on older versions of RealmStudio (and not the latest). Most of the tables are empty though but it should still list each table in the Realm DB.
Thanks,
Clement.

792cddc553a0452ee893e7583735d4e9f0942d6c8404398f612fd5e415e115e19443aa3be112e072f5fc22b7e2471a11ad2924c19413d84d19c32179c1063bd2

**Comment by ironage** [comment]
@clement-ilango sorry, but I am unable to find your attachment when you respond by email to this thread. As far as I can tell, it is a limitation of Github's issue response system. Would you be able to send the file to [realm-help@mongodb.com](mailto:realm-help@mongodb.com) ?

**Comment by clement-ilango** [comment]
I did that.

**Comment by ironage** [comment]
@clement-ilango I have received the files. Thanks for your patience in this process, we will investigate.

**Comment by clement-ilango** [comment]
Hi Nikola Irinchev/James Stone,

This is regarding Jira ticket: RNET-1141<https://urldefense.com/v3/__https:/jira.mongodb.org/browse/RNET-1141__;!!BBM_p3AAtQ!K4dbUDuwEVIqYrz6_SbGnFobOy6NX1wvQ_xMpk-PAd7in42m9QaERrAoKV1TUpidrpL8McOV3kJTbZqaATjhaDEKKtfZ$>. Unfortunately, some business sensitive information has been included in the ticket which must be removed. For this reason, we would like you to do one of the following: -

  1.  Delete or unpublish this RNET-1141<https://urldefense.com/v3/__https:/jira.mongodb.org/browse/RNET-1141__;!!BBM_p3AAtQ!K4dbUDuwEVIqYrz6_SbGnFobOy6NX1wvQ_xMpk-PAd7in42m9QaERrAoKV1TUpidrpL8McOV3kJTbZqaATjhaDEKKtfZ$>.
  2.  Alternatively delete all contents from the ticket. However, if you must, you can update the Jira ticket RNET-1141<https://urldefense.com/v3/__https:/jira.mongodb.org/browse/RNET-1141__;!!BBM_p3AAtQ!K4dbUDuwEVIqYrz6_SbGnFobOy6NX1wvQ_xMpk-PAd7in42m9QaERrAoKV1TUpidrpL8McOV3kJTbZqaATjhaDEKKtfZ$> with content of GitHub issue report that we updated at view it on GitHub<https://urldefense.com/v3/__https:/github.com/realm/realm-dotnet/issues/3592*issuecomment-2089322603__;Iw!!BBM_p3AAtQ!K4dbUDuwEVIqYrz6_SbGnFobOy6NX1wvQ_xMpk-PAd7in42m9QaERrAoKV1TUpidrpL8McOV3kJTbZqaATjhaBlKSW6y$>.

Please note this is an urgent matter for us and we need your cooperation to resolve this issue.

Would it be possible for us to reach out to you via phone?

Thanks,
Clement.

**Comment by nirinchev** [comment]
@clement-ilango I've edited the ticket in Jira to remove the sensitive information. Can you please double check to confirm. If there's more content that needs to be removed, can you please reach out to me via email at nikola.irinchev@mongodb.com.

**Comment by clement-ilango** [comment]
Hi Nikola,
I don’t see that Jira ticket anymore. Could you print it to a PDF and send it to me please?
Thanks,
Clement.

**Comment by ironage** [comment]
We have merged a fix, please try when the next release is available that includes Core v14.10.0

**Comment by clement-ilango** [comment]
Hi James,
Thanks a lot for letting me know that.
Thanks,
Clement.

**Comment by nirinchev** [comment]
Fixed in 12.3.0
